File: app/src/main/java/com/absinthe/kage/connect/tcp/AbstractPacketWriter.java
package com.absinthe.kage.connect.tcp;

import android.util.Log;

import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;

public class AbstractPacketWriter implements IPacketWriter {
    private static final String TAG = AbstractPacketWriter.class.getSimpleName();

    protected static final long DEFAULT_TIMEOUT = 5 * 1000;
    protected long timeout = DEFAULT_TIMEOUT;
    protected LinkedBlockingQueue<Packet> mPacketQueue = new LinkedBlockingQueue<>();
    protected boolean shutdown = false;

    public AbstractPacketWriter(final DataOutputStream out, final KageSocket.ISocketCallback socketCallback) {
        new Thread(() -> {
            while (!shutdown) {
                try {
                    Packet packet = mPacketQueue.poll(timeout, TimeUnit.MILLISECONDS);
                    if (shutdown) {
                        break;
                    }
                    if (null != packet) {
                        writeToStream(out, packet);
                    } else {
                        KageSocket.ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                            if (null != socketCallback) {
                                socketCallback.onWriterIdle();
                            }
                        });
                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    Log.e(TAG, "Send error: " + e.getMessage());
                    KageSocket.ISocketCallback.KageSocketCallbackThreadHandler.getInstance().post(() -> {
                        if (null != socketCallback) {
                            socketCallback.onReadAndWriteError(KageSocket.ISocketCallback.WRITE_ERROR_CODE_CONNECT_UNKNOWN);
                        }
                    });
                }
            }
        }).start();
    }

    protected void writeToStream(DataOutputStream dos, Packet packet) throws IOException {
        String data = packet.getData();
        Log.d(TAG, "Send data: " + data);
        byte[] bArray = data.getBytes(StandardCharsets.UTF_8);
        int sendLen = bArray.length;
        dos.writeInt(sendLen);
        dos.flush();
        dos.write(bArray, 0, sendLen);
        dos.flush();
    }

    @Override
    public void writePacket(Packet packet) {
        try {
            mPacketQueue.put(packet);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void shutdown() {
        this.shutdown = true;
        mPacketQueue.clear();
    }
}

File: app/src/main/java/com/absinthe/kage/connect/tcp/AbstractPacketWriter.kt
package com.absinthe.kage.connect.tcp

import android.util.Log
import com.absinthe.kage.connect.tcp.KageSocket.ISocketCallback
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.DataOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit

open class AbstractPacketWriter(out: DataOutputStream, socketCallback: ISocketCallback?) : IPacketWriter {

    private var timeout = DEFAULT_TIMEOUT
    private var mPacketQueue = LinkedBlockingQueue<Packet?>()
    private var shutdown = false

    @Throws(IOException::class)
    protected fun writeToStream(dos: DataOutputStream, packet: Packet) {
        val data = packet.data
        Log.d(TAG, "Send data: $data")

        val bArray = data!!.toByteArray(StandardCharsets.UTF_8)
        val sendLen = bArray.size

        dos.writeInt(sendLen)
        dos.flush()
        dos.write(bArray, 0, sendLen)
        dos.flush()
    }

    override fun writePacket(packet: Packet?) {
        try {
            mPacketQueue.put(packet)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }

    override fun shutdown() {
        shutdown = true
        mPacketQueue.clear()
    }

    companion object {
        private val TAG = AbstractPacketWriter::class.java.simpleName
        protected const val DEFAULT_TIMEOUT = 5 * 1000.toLong()
    }

    init {
        GlobalScope.launch(Dispatchers.IO) {
            while (!shutdown) {
                try {
                    val packet = mPacketQueue.poll(timeout, TimeUnit.MILLISECONDS)
                    if (shutdown) {
                        break
                    }

                    packet?.let { writeToStream(out, it) }
                            ?: ISocketCallback.KageSocketCallbackThreadHandler.instance?.post {
                                socketCallback?.onWriterIdle()
                            }
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                } catch (e: IOException) {
                    Log.e(TAG, "Send error: " + e.message)
                    ISocketCallback.KageSocketCallbackThreadHandler.instance?.post {
                        socketCallback?.onReadAndWriteError(ISocketCallback.WRITE_ERROR_CODE_CONNECT_UNKNOWN)
                    }
                }
            }
        }
    }
}